### ISSUE TITLE
Target netcoreapp2.1

### ISSUE THREAD

**Issue by Elufimov** [nuke-build/nuke#337]
After update our internal base build library that was target netstandard 2.0 I found that new version of Nuke.Common targets netcoreapp3.0. This force us to stay on 0.22.2 or leave Nuke because we have multiple application with locked sdk on 2.1. What is wrong with netstandard 2.0 that can be used in dotnet core 2.X and and higher?

**Comment by matkoch** [comment]
For several reasons:

- C# 8 language features
- `Microsoft.Build.*` dependencies
- `PackageDownload` feature

Are you planning to stay on 2.1 forever? That doesn’t sound like a good idea.

**Comment by Elufimov** [comment]
dotnet core 2.1 is a LTS release and will have support until 2021 (At least three years from LTS declaration (August 21, 2018)) https://dotnet.microsoft.com/platform/support/policy/dotnet-core. Not everyone can adopt every new version, especially released one month ago.

**Comment by matkoch** [comment]
I suggest compiling your build with 3.0.100 and your other projects with
the older SDk. That’s possible using global.json.

**Comment by Elufimov** [comment]
Yes, it is possible, but it leads to unnecessary pressure to development/qa/devops team. SDK is one of dependencies and this is strange to update it to new major version only for build dependency. This is an open source project and I cant demand anything, however for me it is strange to not support at least a lts versions of sdk for a build library. Be on the cutting edge of new versions crashing backward compatibility is not a good way to go.

**Comment by matkoch** [comment]
What bugs are troubling you in 0.22.3? And what exactly results in
unnecessary pressure?

**Comment by Elufimov** [comment]
> What bugs are troubling you in 0.22.3

Nothing troubles me in 0.22.3, I just wanted to update it as part of adding a new function to our internal build framework use Nuke.   

> And what exactly results in unnecessary pressure?

Work that need to be done by people. This is a technical debt, that will take time from teams, and time is money. 
As I said we can stay on 0.22.3, but this is disturbing that this project can drop support o sdk that will be supported by Microsoft within another 2 years.

**Comment by matkoch** [comment]
It is literally creating two global.json files? I don’t understand how this
can be so complicated.

**Comment by Elufimov** [comment]
As I can remember dotnet cli looking for global.json in current working directory. I need to rewrite default build.ps1/sh files and call nuke not from root directory because IDE will use the root one. But this need to be tested and looking not that simple. Or I don't really understand you.

**Comment by matkoch** [comment]
I'm sorry, but I have a little issue with the sentiment here. You're saying that 0.22.3 works for you without issues. If there really was a bug, that would be a different discussion, as we still can do hotfix releases. Okay. Now you want to integrate new features from 0.23.0. Alright, the same goes for me about C# 8 language features (but also because `Microsoft.Build` now requires `netcoreapp3.0`). So I've done my part of upgrading everything, also having in mind that multi-targeting isn't worth the effort (and kinda kills the argument for C# 8 syntax), because on user-side it's possible to compile the build project with 3.x and the rest of the solution with whatever someone needs. If this is too hard for you, you can just wait and schedule it for whatever time works for you. There is apparently no bugfix-reason for you to upgrade anything right now.

In other words, lets say 0.22.3 is the LTS for this project. Although I don't really offer a service here.

**Comment by Elufimov** [comment]
It is you right to do what ever you want, and I do not mean to oblige with some sort of service. All I want to see why you discard huge part of actual skds. And if you clearly do not want to make a conversations about how to overcome using new version of nuke on current lts sdk I will only add this. All c#8 features will work only on dotnet core 3.0 and you will need drop net472, or use portion of features like pointed here https://stackoverflow.com/a/57020770 As i see you use Microsoft.Build.*  16.0.461 and according to nuget.org it require .NETCoreApp 2.1 not 3.0. You can close this issue. If I will find bug in current LTS (0.22.3) I will create a separate issue.

**Comment by matkoch** [comment]
lets try https://github.com/nuke-build/common/tree/feature/netcoreapp2.1

**Comment by Elufimov** [comment]
Thx for this branch! I will certainly give it a try on monday!

**Comment by dasMulli** [comment]
> SDK is one of dependencies and this is strange to update it to new major version only for build dependency. 

Pinning an SDK to an older version is tricky and can also have some unintended side-effects. For example, there are two currently supported versions for building .NET Core 2.1 apps (2.1.6xx and 2.1.8xx - both are latest versinos) that only differ in the MSBuild, NuGet, compiler and test platform versions being integrated - one for VS 2017 and one for VS 2019. In theory it is possible to use the 2.1.6xx targets from an msbuild installation from VS 2019 but not the other way around. It less interesting when there is a minor version difference between the tools used in the SDK and the hosting msbuild+tools instance that is used (e.g. Visual Studio would have a newer compiler with a bug fix but the CI build would use an older one).

The latest SDKs also support building for .NET Core versions down to 1.0. While moving to the 3.0 SDK can technically have breaking changes (such as NuGet removing deprecated features or the -o arguments being interpreted differently), it does not affect which support policies (e.g. LTS) apply to your main project.

Maintaining a netcoreapp2.1 is possible but I think as an open source project that does not require you to update your main application (only the build), Nuke is free to decide on how much (free!) effort they spend on maintaining backwards compatibility. Also don't forget that for now, 3.1 (going to become the next LTS) is only weeks away.

If you care about maintaining perfect build reproducibility to avoid QA issues, maybe you shouldn't try update Nuke or even your Visual Studio version. (that's why MS also supports the RTM minor versions (15.0.*, 16.0.*) of VS)

**Comment by Elufimov** [comment]
@matkoch do you have a nightly builds? 
@dasMulli 
>compiler and test platform versions being integrated - one for VS 2017 and one for VS 2019

CI in docker container that have only dotnet core sdk allows do not think about difference in different versions of VS. Also one can use rider or even vs code and does not have VS at all.

> Maintaining a netcoreapp2.1 is possible but I think as an open source project that does not require you to update your main application (only the build)

There is a trick that nuke is in fact a part of the main solution and, in normal circumstances, builds under solution wide sdk. And this is the main selling point of this build framework. It just part of solution, no need of separate tools, or plugins for work with it. 

> , Nuke is free to decide on how much (free!) effort they spend on maintaining backwards compatibility. 

And I do not denied project to do so. However I really like Nuke.build and want see my case as not some corner case. Every time when some project cut off some platforms it cut off some portion of users. I want more people use Nuke, not less. If there is a chance save some platforms why not point on it? 

>, maybe you shouldn't try update Nuke or even your Visual Studio version

Personally I use Rider at work and do not have VS at all.

**Comment by matkoch** [comment]
Closing due to inactivity.

**Comment by Elufimov** [comment]
@matkoch This is still relevant to at least to me. How can I help? Two days ago I successfully (with tests run) compiled this repository as multi target of netcoreapp2.1;netcoreapp3.0;net472 without any code changes. I did not try results under limiting environment though (target 2.1 with global 2.1 and target 3.0 with global 3.0 and target 3.1 with global 3.1).

**Comment by Hawxy** [comment]
I really don't see how this is an issue. 

Per https://docs.microsoft.com/en-us/dotnet/core/versions/selection The 3.1 SDK can build anything back to .NET Core 1.0. Just install the latest SDK in your environments and everything else will continue to build fine. At my work we have the 3.1 SDK installed on our build machines and use it for everything back to .NET Core 2.0 without any issue or noticeable change. 

You're kind of an extreme edge-case whereby you've locked yourself into a specific SDK version because, as far as I can tell, your development pipeline isn't flexible enough to spend 5 minutes installing a new one. 

> CI in docker container

If your CI is a docker container, isn't it just a one-line change from `2.1` to `3.1` for the SDK version?

> This is a technical debt, that will take time from teams, and time is money.

Get rid of your global.json's (therefore always building with the latest SDK) and it's no longer technical debt. 

<blockquote><img src="https://docs.microsoft.com/en-us/media/logos/logo-ms-social.png" width="48" align="right"><div><strong><a href="https://docs.microsoft.com/en-us/dotnet/core/versions/selection">Select which .NET Core version to use - .NET Core</a></strong></div><div>Learn how .NET Core automatically finds and chooses runtime versions for your program. Additionally, this article teaches you how to force a specific version.</div></blockquote>

**Comment by Elufimov** [comment]
@Hawxy Your tone is so uncompromisable... 
> as far as I can tell, your development pipeline isn't flexible enough to spend 5 minutes installing a new one.

As far as I can tell, your development pipeline is too careless in term of possible regression. This is how I can see this comment of yours from "different extreme edge-case".

> If your CI is a docker container, isn't it just a one-line change from 2.1 to 3.1 for the SDK version?

If we are talking about just building a new runtime image than yes, if we consider to run a performance and regressions tests of our applications than no.

> Get rid of your global.json's (therefore always building with the latest SDK) and it's no longer technical debt.

And than start helping developers and aqas with weird bugs on older projects. No thx.

As far for this issue, we use conditional dependencies in our build project that helps us target to 2.1 and 3.0. There are not so many api difference between 0.22.2 and 0.24.1 versions.

```csharp
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<TargetFrameworks>netcoreapp2.1;netcoreapp3.1</TargetFrameworks>
		<IsPackable>True</IsPackable>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Nuke.Docker" Version="0.4.0" />
	</ItemGroup>

	<ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp2.1'">
		<PackageReference Include="GitVersion.CommandLine.DotNetCore" Version="5.0.0" />
		<PackageReference Include="Nuke.Common" Version="0.22.2" />
	</ItemGroup>

	<ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.1'">
		<PackageReference Include="Nuke.Common" Version="0.24.1" />
		<PackageDownload Include="GitVersion.Tool" Version="[5.1.3]" />
	</ItemGroup>

</Project>
```  

This solution is fine to us.

**Comment by RemyDuijkeren** [comment]
I also can't upgrade nuke anymore since version 0.22.2. I use NuGet libs in my build that unfortunately require my build project to target .NET 4.6.2 (Dynamics 365). This forces me to solve it in different way, like moving this dependency into a separate console app, which I am trying to prevent.

Limiting Nuke.build only to .NET 4.7.2 and .NET Core 3.0 seems to me very strict. Adding target .NET Standard 2.0 would help.

**Comment by matkoch** [comment]
I reduced target framework to `netcoreapp2.1` with 0.24.4

**Comment by matkoch** [comment]
@remyvd could you star the project?

**Comment by matkoch** [comment]
@Elufimov was this correctly resolved for your case?
